src/xtgeo/io/tsurf/_tsurf_reader.py

```diff
@@ -0,0 +1,799 @@
+            continue
+
+        # -----------------------
+        # Handle unknown keywords
+        # -----------------------
+        # The TSurf file format handles many more keywords and attributes
+        # than those captured in this parser. Could potentially issue a warning
+        # instead of an error and continue. But if the section
+        # continues over several lines, it is difficult to know where it
+        # ends and where parsing can resume.
+
+        raise ValueError(
+            f"\nIn file {filepath_errmsg}:\n"
+            "The file contains an invalid line which is not recognized \n"
+            "as a section identifier (first line of a section).\n"
+            "This may be either an error, or a valid TSurf section identifier\n"
+            "that is not (yet) handled by the file parser.\n"
+            f"Failing line: '{line}'",
+        )
+
+    # Ensure mandatory sections were read
+    if not header_section_read:
+        raise ValueError(
+            f"\nIn file {filepath_errmsg}:\nMissing mandatory 'HEADER' section.\n"
+        )
+    if not tface_section_read:
+        raise ValueError(
+            f"\nIn file {filepath_errmsg}:\nMissing mandatory 'TFACE' section.\n"
+        )
+
+    tsurf_data = _create_tsurf_data(header_name, coord_sys_data, vertices, triangles)
+    _validate_triangulation_data(tsurf_data.vertices, tsurf_data.triangles)
+    return tsurf_data
+
+
+def _validate_file_path(file: Path, filepath_errmsg: str) -> None:
+    """Validate that file path exists and has correct extension.
+
+    Args:
+        file: Path object to validate
+        filepath_errmsg: Error context for meaningful error messages
+
+    Raises:
+        FileNotFoundError: If file doesn't exist or isn't a regular file
+        ValueError: If file doesn't have .ts extension
+    """
+    if not file.exists():
+        raise FileNotFoundError(f"\nFile {filepath_errmsg}:\nThe file does not exist.")
+
+    if not file.is_file():
+        raise FileNotFoundError(
+            f"\nFile {filepath_errmsg}:\n"
+            + "The file is not a regular file. "
+            + "It may be a directory or a special file type."
+        )
+
+    if file.suffix != ".ts":
+        raise ValueError(
+            f"\nFile {filepath_errmsg}:\n"
+            "The file is not a TSurf file. Expected '.ts' extension."
+        )
+
+
+@contextmanager
+def _get_text_stream(
+    file: FileLike, encoding: str
+) -> Generator[Union[TextIOWrapper, StringIO], None, None]:
+    """Context manager that yields a text stream.
+
+    Handles different input types:
+    - str: converts to Path and opens file with specified encoding
+    - Path objects: opens file with specified encoding
+    - BytesIO: wraps with TextIOWrapper with specified encoding
+    - StringIO: yields directly (already text-based)
+    """
+    if isinstance(file, (str, Path)):
+        file_path = Path(file) if isinstance(file, str) else file
```

Comment:
```suggestion
        file_path = Path(file)
```

`Path()`ing a Path is totally fine, skip the condition